### PR TITLE
feature/pod-nft

### LINKED ISSUE #62: Add a "Paid On Delivery"(POD) trade function
Blocked by  #10 
In the Bank pallet, add a paid-on-delivery (POD) mechanism

Users can transfer an NFT to another user and define a price in Native currency. A transfer request is created against the target user and an ID is assigned. A Fixed fee is required from the sender and is paid to the treasury to prevent spamming.

The receiving user can accept the delivery through a separate extrinsic, where they will pay the sender the correct amount of funds + a tip of their choice. in exchange for receiving the NFT in the delivery. They can also reject the POD where the request is removed and the NFT unlocked. The Fee is not refunded.
 
The NFT is locked while it is being PODed. (cannot be POD and sold on Auction). Locked NFT can not be PODed.

If a request is not accepted after the "expiry" period, it is removed and the NFT is unlocked. the fee is not refunded.

The sender can also cancel a POD. The fee is not refunded.

### PR DESCRIPTION
**Pull request by IRISxiaowang**
Close #62

Added buy_nft() and sell_nft() in pallet-bank.
Added trait ManageNfts
Added ensure_not_role() in trait ManageRoles

### CODE REVIEW

**Comment by IRISxiaowang** [comment]
Moved pod from pallet-bank to pallet-nft

File: pallets/pallet-nft/src/lib.rs [hunk 1]
```diff
@@ -75,14 +126,32 @@ pub mod module {
...
+		/// Indicates that someone received an Nft with a certain price.
+		NftDelivered {
+			seller: T::AccountId,
+			buyer: T::AccountId,
+			nft_id: NftId,
+			price: T::Balance,
+			tips: T::Balance,
+		},
+		/// Indicates that the creator canceled the nft on POD.
+		NftCanceled { nft_id: NftId, reason: UnlockReason },
```
Comment (syan095) [inline pallets/pallet-nft/src/lib.rs]: maybe call this cancel reason

File: pallets/pallet-nft/src/lib.rs [hunk 2]
```diff
@@ -95,10 +164,32 @@ pub mod module {
 	#[pallet::getter(fn nfts)]
 	pub type Nfts<T: Config> = StorageMap<_, Blake2_128Concat, NftId, NftData>;
 
+	/// Stores the pricing information for NFTs awaiting trade.
+	#[pallet::storage]
+	#[pallet::getter(fn pending_trade_nfts)]
+	pub type PendingPodNfts<T: Config> =
+		StorageMap<_, Blake2_128Concat, PodId, PodInfo<T::AccountId, T::Balance>>;
+
+	/// Stores the users' nft ids that will have their nft unlocked at a block.
```
Comment (syan095) [inline pallets/pallet-nft/src/lib.rs]: this should be about expiry for POD pending delivery, not unlock

File: pallets/pallet-nft/src/lib.rs [hunk 3]
```diff
@@ -41,27 +64,53 @@ pub mod module {
...
 		Unauthorised,
-		/// The caller's role does not permit this action.
-		IncorrectRole,
 		/// The specified NFT ID does not exist.
 		InvalidNftId,
 		/// The provided data exceeds the allowed size limit.
 		DataTooLarge,
 		/// The provided file name exceeds the maximum permitted length.
 		FileNameTooLarge,
+		/// The nft id does not for POD.
```
Comment (syan095) [inline pallets/pallet-nft/src/lib.rs]: clarify this

File: pallets/pallet-nft/src/lib.rs [hunk 4]
```diff
@@ -41,27 +64,53 @@ pub mod module {
...
+		NftNotForPod,
+		/// The nft id does not exist.
+		NftNotExist,
```
Comment (syan095) [inline pallets/pallet-nft/src/lib.rs]: NftDoesNotExist

File: pallets/pallet-nft/src/lib.rs [hunk 5]
```diff
@@ -75,14 +124,32 @@ pub mod module {
 		NftTransferred { from: T::AccountId, to: T::AccountId, nft_id: NftId },
 		/// Emitted when a new NFT is requested to be minted, pending audit.
 		NFTPending { nft_id: NftId, file_name: Vec<u8> },
-		/// Emitted when an NFT creation is rejected by the auditor.
+		/// Emitted when an NFT creation is rejected by the auditor or the receiver rejected the
```
Comment (syan095) [inline pallets/pallet-nft/src/lib.rs:125]: Should use a different event for when POD is rejected.
It represents a different scenario from when NFT MInt is rejected.

File: pallets/pallet-nft/src/lib.rs [hunk 6]
```diff
@@ -75,14 +124,32 @@ pub mod module {
...
+		/// Indicates that someone received an Nft with a certain price.
+		NftDelivered {
+			seller: T::AccountId,
+			buyer: T::AccountId,
+			nft_id: NftId,
+			price: T::Balance,
+			tips: T::Balance,
+		},
+		/// Indicates that the creator canceled the nft on POD.
+		CancelReason { nft_id: NftId, reason: UnlockReason },
```
Comment (syan095) [inline pallets/pallet-nft/src/lib.rs]: Rename this event

File: pallets/pallet-nft/src/lib.rs [hunk 7]
```diff
@@ -95,10 +162,32 @@ pub mod module {
...
+
 	#[pallet::pallet]
 	#[pallet::without_storage_info]
 	pub struct Pallet<T>(_);
 
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_finalize(block_number: BlockNumberFor<T>) {
+			// Unlock nfts that are due.
+			UnlockNft::<T>::take(block_number).into_iter().for_each(|(pod_id, nft_id)| {
```
Comment (syan095) [inline pallets/pallet-nft/src/lib.rs]: Should rename this storage to something like PodExpiry
Avoid using "unlock/lock" for this, use "expire", cancel or "completed".

i.e.: "Ignore the cancel result - pod could be cancelled or completed prior to the expiry."

File: runtime/src/lib.rs [hunk 8]
```diff
@@ -283,8 +283,12 @@ impl pallet_nft::Config for Runtime {
 	type RuntimeEvent = RuntimeEvent;
 	type WeightInfo = ();
 	type RoleManager = Roles;
+	type Balance = Balance;
+	type Bank = Bank;
 	type EnsureGovernance = traits::SuccessOrigin<Runtime>;
 	type MaxSize = ConstU32<1000>;
+	type Fee = ConstU128<CENT>;
```
Comment (syan095) [inline runtime/src/lib.rs]: should allow higher MaxSize
and also use higher Fee (at least a dollar)
Rename Fee to something more clear - PodFee

**Review by syan095 — CHANGES_REQUESTED** [review]
Should add 3 Integration tests:
* created a POD and receiver accepted
* Receiver rejected
* Sender cancelled

**Review by syan095 — APPROVED** [review]

**Pull request merged**